### PR TITLE
ARROW-1992: [C++/Python] Fix segfault when string to categorical empty string array

### PR DESCRIPTION
**Pull request by Licht-T**
This closes [ARROW-1992](https://issues.apache.org/jira/browse/ARROW-1992).

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/hash.cc [hunk 1]
```diff
@@ -406,20 +406,29 @@ class HashTableKernel<Type, Action, enable_if_binary<Type>> : public HashTable {
...
     auto action = static_cast<Action*>(this);
     RETURN_NOT_OK(action->Reserve(arr.length));
 
 #define HASH_INNER_LOOP()                                                           \
   const int32_t position = offsets[i];                                              \
   const int32_t length = offsets[i + 1] - position;                                 \
-  const uint8_t* value = data + position;                                           \
+  const uint8_t* value = data;                                                      \
+  if (value != &empty_value) {                                                      \
+    value += position;                                                              \
```
Comment (wesm) [inline cpp/src/arrow/compute/kernels/hash.cc]: When this case comes up, `position` is always going to be 0, so we could maybe handle this by substituting a dummy static buffer for `data` above.

**Comment by Licht-T** [comment]
Thanks @wesm, fixed!

**Review by wesm — APPROVED** [review]
+1, thanks @Licht-T!

**Review by xhochy — APPROVED** [review]
+1, failure is only due to Thrift

**Pull request closed without merging**